File: src/main/persistence/learnings.ts
import { randomUUID } from "node:crypto";
import type Database from "better-sqlite3";
import { RecordNotFoundError } from "./errors.js";
import type { Learning } from "../../shared/types.js";

interface LearningRow {
  id: string;
  project_id: string;
  kind: "pitfall" | "convention" | "command";
  summary: string;
  evidence_session_id: string | null;
  evidence_event_id: string | null;
  verified: number;
  hits: number;
  created_at: string;
  last_seen_at: string;
}

export interface InsertLearningInput {
  id?: string;
  projectId: string;
  kind: "pitfall" | "convention" | "command";
  summary: string;
  evidenceSessionId?: string | null;
  evidenceEventId?: string | null;
}

function learningRowToSummary(row: LearningRow): Learning {
  return {
    id: row.id,
    projectId: row.project_id,
    kind: row.kind,
    summary: row.summary,
    evidenceSessionId: row.evidence_session_id,
    evidenceEventId: row.evidence_event_id,
    verified: row.verified === 1,
    hits: row.hits,
    createdAt: row.created_at,
    lastSeenAt: row.last_seen_at
  };
}

export function insertLearning(connection: Database.Database, input: InsertLearningInput): Learning {
  const id = input.id ?? randomUUID();
  const now = new Date().toISOString();
  connection
    .prepare(
      `INSERT INTO learnings (id, project_id, kind, summary, evidence_session_id, evidence_event_id, created_at, last_seen_at)
       VALUES (?, ?, ?, ?, ?, ?, ?, ?)`
    )
    .run(
      id,
      input.projectId,
      input.kind,
      input.summary,
      input.evidenceSessionId ?? null,
      input.evidenceEventId ?? null,
      now,
      now
    );
  const row = connection.prepare("SELECT * FROM learnings WHERE id = ?").get(id) as LearningRow;
  return learningRowToSummary(row);
}

export function listLearnings(
  connection: Database.Database,
  projectId: string,
  limit = 50
): Learning[] {
  const rows = connection
    .prepare(
      `SELECT * FROM learnings WHERE project_id = ?
       ORDER BY verified DESC, hits DESC, last_seen_at DESC
       LIMIT ?`
    )
    .all(projectId, limit) as LearningRow[];
  return rows.map(learningRowToSummary);
}

export function updateLearning(
  connection: Database.Database,
  input: { id: string; summary?: string; verified?: boolean }
): Learning {
  const fragments: string[] = [];
  const values: unknown[] = [];
  if (typeof input.summary === "string") {
    fragments.push("summary = ?");
    values.push(input.summary);
  }
  if (typeof input.verified === "boolean") {
    fragments.push("verified = ?");
    values.push(input.verified ? 1 : 0);
  }
  if (fragments.length === 0) {
    const existing = connection.prepare("SELECT * FROM learnings WHERE id = ?").get(input.id) as LearningRow | undefined;
    if (!existing) throw new RecordNotFoundError("learning", input.id);
    return learningRowToSummary(existing);
  }
  fragments.push("last_seen_at = ?");
  values.push(new Date().toISOString());
  values.push(input.id);
  const result = connection
    .prepare(`UPDATE learnings SET ${fragments.join(", ")} WHERE id = ?`)
    .run(...values);
  if (result.changes === 0) {
    throw new RecordNotFoundError("learning", input.id);
  }
  const row = connection.prepare("SELECT * FROM learnings WHERE id = ?").get(input.id) as LearningRow;
  return learningRowToSummary(row);
}

export function deleteLearning(connection: Database.Database, id: string): void {
  connection.prepare("DELETE FROM learnings WHERE id = ?").run(id);
}

/**
 * Ranked full-text search over events.message via the FTS5 sidecar. Returns
 * the matching session id, event id, an FTS5 snippet (with `<b>` markers
 * around the matched tokens), and the raw rank (lower = more relevant).
 *
 * Free-text input is quoted and escaped so FTS5 operator syntax doesn't leak
 * — `"`, `*`, `(`, `AND`/`OR`/`NEAR` etc. are no longer special. Use
 * {@link searchEventsRaw} when you actually want operator syntax (CLI usage,
 * power-user palette).
 */
export function searchEvents(
  connection: Database.Database,
  query: string,
  limit = 50
): Array<{ sessionId: string; eventId: string; snippet: string; rank: number }> {
  return searchEventsRaw(connection, escapeFts5(query), limit);
}

/** Wrap user input so FTS5 treats it as a literal phrase, escaping any
 *  embedded double quotes. Whitespace inside the phrase still tokenises. */
function escapeFts5(query: string): string {
  const trimmed = query.trim();
  if (!trimmed) return "";
  return `"${trimmed.replace(/"/g, '""')}"`;
}

export function searchEventsRaw(
  connection: Database.Database,
  query: string,
  limit = 50
): Array<{ sessionId: string; eventId: string; snippet: string; rank: number }> {
  if (!query.trim()) return [];
  const rows = connection
    .prepare(
      `SELECT events.session_id AS sessionId,
              events.id AS eventId,
              snippet(events_fts, 0, '<b>', '</b>', '…', 12) AS snippet,
              events_fts.rank AS rank
       FROM events_fts
       JOIN events ON events.rowid = events_fts.rowid
       WHERE events_fts MATCH ?
       ORDER BY events_fts.rank
       LIMIT ?`
    )
    .all(query, limit) as Array<{
    sessionId: string;
    eventId: string;
    snippet: string;
    rank: number;
  }>;
  return rows;
}

File: src-tauri/src/persistence/learnings.rs
use rusqlite::{Connection, Row};
use serde::Serialize;
use specta::Type;
use uuid::Uuid;

use super::prepared::prepared;
use super::time::now_iso;
use crate::error::{ArgmaxError, ArgmaxResult};

#[derive(Debug, Clone, PartialEq)]
pub struct InsertLearningInput {
    pub id: Option<String>,
    pub project_id: String,
    pub kind: String,
    pub summary: String,
    pub evidence_session_id: Option<String>,
    pub evidence_event_id: Option<String>,
}

#[derive(Debug, Clone, PartialEq)]
pub struct UpdateLearningInput {
    pub id: String,
    pub summary: Option<String>,
    pub verified: Option<bool>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Type)]
#[serde(rename_all = "camelCase")]
pub struct Learning {
    pub id: String,
    pub project_id: String,
    pub kind: String,
    pub summary: String,
    pub evidence_session_id: Option<String>,
    pub evidence_event_id: Option<String>,
    pub verified: bool,
    pub hits: i64,
    pub created_at: String,
    pub last_seen_at: String,
}

#[derive(Debug, Clone, PartialEq, Serialize, Type)]
#[serde(rename_all = "camelCase")]
pub struct EventSearchResult {
    pub session_id: String,
    pub event_id: String,
    pub snippet: String,
    pub rank: f64,
}

pub fn insert_learning(
    connection: &Connection,
    input: &InsertLearningInput,
) -> ArgmaxResult<Learning> {
    let id = input
        .id
        .clone()
        .unwrap_or_else(|| Uuid::new_v4().to_string());
    let now = now_iso();
    let mut statement = prepared(
        connection,
        r#"
        INSERT INTO learnings (id, project_id, kind, summary, evidence_session_id, evidence_event_id, created_at, last_seen_at)
        VALUES (?, ?, ?, ?, ?, ?, ?, ?)
        "#,
    )
    .map_err(sqlite_error)?;
    statement
        .execute((
            id.as_str(),
            input.project_id.as_str(),
            input.kind.as_str(),
            input.summary.as_str(),
            input.evidence_session_id.as_deref(),
            input.evidence_event_id.as_deref(),
            now.as_str(),
            now.as_str(),
        ))
        .map_err(sqlite_error)?;
    find_learning_by_id(connection, &id)
}

pub fn list_learnings(
    connection: &Connection,
    project_id: &str,
    limit: usize,
) -> ArgmaxResult<Vec<Learning>> {
    let mut statement = prepared(
        connection,
        r#"
        SELECT * FROM learnings WHERE project_id = ?
        ORDER BY verified DESC, hits DESC, last_seen_at DESC
        LIMIT ?
        "#,
    )
    .map_err(sqlite_error)?;
    let rows = statement
        .query_map((project_id, limit as i64), row_to_learning)
        .map_err(sqlite_error)?
        .collect::<Result<Vec<_>, _>>()
        .map_err(sqlite_error)?;
    Ok(rows)
}

pub fn update_learning(
    connection: &Connection,
    input: &UpdateLearningInput,
) -> ArgmaxResult<Learning> {
    match (&input.summary, input.verified) {
        (None, None) => find_learning_by_id(connection, &input.id),
        (Some(summary), Some(verified)) => {
            let mut statement = prepared(
                connection,
                "UPDATE learnings SET summary = ?, verified = ?, last_seen_at = ? WHERE id = ?",
            )
            .map_err(sqlite_error)?;
            update_learning_row(
                &mut statement,
                (
                    summary.as_str(),
                    bool_to_i64(verified),
                    now_iso(),
                    input.id.as_str(),
                ),
                &input.id,
            )?;
            find_learning_by_id(connection, &input.id)
        }
        (Some(summary), None) => {
            let mut statement = prepared(
                connection,
                "UPDATE learnings SET summary = ?, last_seen_at = ? WHERE id = ?",
            )
            .map_err(sqlite_error)?;
            update_learning_row(
                &mut statement,
                (summary.as_str(), now_iso(), input.id.as_str()),
                &input.id,
            )?;
            find_learning_by_id(connection, &input.id)
        }
        (None, Some(verified)) => {
            let mut statement = prepared(
                connection,
                "UPDATE learnings SET verified = ?, last_seen_at = ? WHERE id = ?",
            )
            .map_err(sqlite_error)?;
            update_learning_row(
                &mut statement,
                (bool_to_i64(verified), now_iso(), input.id.as_str()),
                &input.id,
            )?;
            find_learning_by_id(connection, &input.id)
        }
    }
}

pub fn delete_learning(connection: &Connection, id: &str) -> ArgmaxResult<()> {
    let mut statement =
        prepared(connection, "DELETE FROM learnings WHERE id = ?").map_err(sqlite_error)?;
    statement.execute([id]).map_err(sqlite_error)?;
    Ok(())
}

pub fn search_events(
    connection: &Connection,
    query: &str,
    limit: usize,
) -> ArgmaxResult<Vec<EventSearchResult>> {
    search_events_raw(connection, &build_fts_prefix_query(query), limit)
}

pub fn search_events_raw(
    connection: &Connection,
    query: &str,
    limit: usize,
) -> ArgmaxResult<Vec<EventSearchResult>> {
    if query.trim().is_empty() {
        return Ok(Vec::new());
    }
    let mut statement = prepared(
        connection,
        r#"
        SELECT events.session_id AS session_id,
               events.id AS event_id,
               snippet(events_fts, 0, '<b>', '</b>', '...', 12) AS snippet,
               events_fts.rank AS rank
        FROM events_fts
        JOIN events ON events.rowid = events_fts.rowid
        WHERE events_fts MATCH ?
        ORDER BY events_fts.rank
        LIMIT ?
        "#,
    )
    .map_err(sqlite_error)?;
    let rows = statement
        .query_map((query, limit as i64), |row| {
            Ok(EventSearchResult {
                session_id: row.get("session_id")?,
                event_id: row.get("event_id")?,
                snippet: row.get("snippet")?,
                rank: row.get("rank")?,
            })
        })
        .map_err(sqlite_error)?
        .collect::<Result<Vec<_>, _>>()
        .map_err(sqlite_error)?;
    Ok(rows)
}

fn find_learning_by_id(connection: &Connection, id: &str) -> ArgmaxResult<Learning> {
    let mut statement =
        prepared(connection, "SELECT * FROM learnings WHERE id = ?").map_err(sqlite_error)?;
    match statement.query_row([id], row_to_learning) {
        Ok(learning) => Ok(learning),
        Err(rusqlite::Error::QueryReturnedNoRows) => {
            Err(ArgmaxError::record_not_found("learning", id))
        }
        Err(error) => Err(sqlite_error(error)),
    }
}

fn row_to_learning(row: &Row<'_>) -> rusqlite::Result<Learning> {
    Ok(Learning {
        id: row.get("id")?,
        project_id: row.get("project_id")?,
        kind: row.get("kind")?,
        summary: row.get("summary")?,
        evidence_session_id: row.get("evidence_session_id")?,
        evidence_event_id: row.get("evidence_event_id")?,
        verified: row.get::<_, i64>("verified")? == 1,
        hits: row.get("hits")?,
        created_at: row.get("created_at")?,
        last_seen_at: row.get("last_seen_at")?,
    })
}

fn update_learning_row<P>(
    statement: &mut rusqlite::CachedStatement<'_>,
    params: P,
    id: &str,
) -> ArgmaxResult<()>
where
    P: rusqlite::Params,
{
    let changes = statement.execute(params).map_err(sqlite_error)?;
    if changes == 0 {
        return Err(ArgmaxError::record_not_found("learning", id));
    }
    Ok(())
}

fn build_fts_prefix_query(query: &str) -> String {
    let tokens = query
        .split(|character: char| !character.is_ascii_alphanumeric() && character != '_')
        .filter(|token| !token.is_empty())
        .collect::<Vec<_>>();
    if tokens.is_empty() {
        return String::new();
    }
    let phrase = tokens.join(" ");
    let prefixed = tokens
        .iter()
        .filter(|token| !is_fts_operator_token(token))
        .map(|token| format!("{token}*"))
        .collect::<Vec<_>>()
        .join(" ");
    if prefixed.is_empty() {
        return format!("\"{phrase}\"");
    }
    format!("\"{phrase}\" OR ({prefixed})")
}

fn is_fts_operator_token(token: &str) -> bool {
    token.eq_ignore_ascii_case("AND")
        || token.eq_ignore_ascii_case("OR")
        || token.eq_ignore_ascii_case("NOT")
        || token.eq_ignore_ascii_case("NEAR")
}

fn bool_to_i64(value: bool) -> i64 {
    if value {
        1
    } else {
        0
    }
}

fn sqlite_error(error: rusqlite::Error) -> ArgmaxError {
    ArgmaxError::service("SQLITE", error.to_string())
}
